gpflow/models/model.py

```diff
@@ -283,11 +239,31 @@ def predict_y(
         return self.likelihood.predict_mean_and_var(f_mean, f_var)
 
     def predict_log_density(
-        self, data: RegressionData, full_cov: bool = False, full_output_cov: bool = False
+        self, data: RegressionData, full_cov: bool = False, full_output_cov: bool = False,
     ):
         """
         Compute the log density of the data at the new data points.
         """
         X, Y = data
         f_mean, f_var = self.predict_f(X, full_cov=full_cov, full_output_cov=full_output_cov)
         return self.likelihood.predict_density(f_mean, f_var, Y)
+
+
+class InternalDataGPModel(GPModel, InternalDataTrainingInterface):
+    """
+    Base class for models that encapsulate their data and store it as
+    self.data; training_loss does not take any arguments.
+    """
+
+    def training_loss(self):
+        return -self.maximum_a_posteriori_objective()
+
+
+class ExternalDataGPModel(GPModel, ExternalDataTrainingInterface):
+    """
+    Base class for GP models that do not encapsulate the data; training_loss takes
+    data as argument.
+    """
+
+    def training_loss(self, data):
+        return -self.maximum_a_posteriori_objective(data)
```

Comment:
you could completely decouple these from the `GPModel`, which may help with keeping the hierarchy simple, using
```python3
class InternalDataGPModel(GPModel, InternalDataTrainingInterface):
    @abstractmethod
    def maximum_a_posteriori_objective(self):
        pass

    def training_loss(self):
        return -self.maximum_a_posteriori_objective()
```

Reply:
this would also make the mixin work with `GPModel` and `BayesianModel` without assuming either

Reply:
they could even be merged with `InternalDataTrainingInterface` and `ExternalDataTrainingInterface` then

Reply:
The training_loss / maximum_a_posteriori_objective should go with BayesianModel not GPModel.

Reply:
ok. that's fine. it's just that we have the option to do either

Reply:
im not entirely sure what that buys you, except that once again ppl need to remember to add the mixin for their new derived model (which im not so keen on). lets discuss over zoom